ga4gh/versioning.py

```diff
@@ -0,0 +1,152 @@
+    DEVELOPMENT = "DEVELOPMENT"
+    ACTIVE = "ACTIVE"
+    DEACTIVATED = "DEACTIVATED"
+
+
+class Version(object):
+    """
+    A major/minor/revision version tag
+    """
+    @classmethod
+    def parseString(cls, versionString):
+        versions = versionString.strip('vV').split('.')
+        return Version(*versions)
+
+    def __init__(self, major, minor, revision):
+        self.major = major
+        self.minor = minor
+        self.revision = revision
+
+    def __cmp__(self, other):
+        majorCmp = cmp(self.major, other.major)
+        if majorCmp != 0:
+            return majorCmp
+        minorCmp = cmp(self.minor, other.minor)
+        if minorCmp != 0:
+            return minorCmp
+        else:
+            return cmp(self.revision, other.revision)
+
+    def __hash__(self):
+        return hash((self.major, self.minor, self.revision))
+
+    def __eq__(self, other):
+        return (self.major == other.major and
+                self.minor == other.minor and
+                self.revision == other.revision)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class RouteGroup(object):
+    """
+    A group that a route belongs to
+    """
+    def __init__(self, lifecycleStage=Lifecycle.DEVELOPMENT,
+                 enabledVersion=None, disabledVersion=None):
+        self.lifecycleStage = lifecycleStage
+        if enabledVersion is not None:
+            self.enabledVersion = Version.parseString(enabledVersion)
+        else:
+            self.enabledVersion = None
+        if disabledVersion is not None:
+            self.disabledVersion = Version.parseString(disabledVersion)
+        else:
+            self.disabledVersion = None
+
+
+class VersioningManager(object):
+
+    # TODO externalize data fields to yaml file?
+    supportedVersionsData = ['0.5.1']
+
+    groupData = {
+        'V_0.5.1_RELEASE': {
+            'enabledVersion': '0.5.1',
+        },
+    }
+
+    def __init__(self, app):
+        self.groups = self._initGroups()
+        self.supportedVersions = self._initSupportedVersions()
+        self.app = app
+
+    def _initSupportedVersions(self):
+        supportedVersions = set()
+        for supportedVersion in self.supportedVersionsData:
+            version = Version.parseString(supportedVersion)
+            supportedVersions.add(version)
+        return supportedVersions
+
+    def _initGroups(self):
+        groups = {}
+        for groupName, groupDataDict in self.groupData.items():
+            group = RouteGroup(**groupDataDict)
+            groups[groupName] = group
+        return groups
+
+    def policy(self, routeGroup):
+        def inner(func):
+            @functools.wraps(func)
+            def decoratedFunction(version):
+                if self.isAccessAllowed(routeGroup, version):
+                    return func()
+                else:
+                    raise Exception('Access denied')  # TODO custom exception
```

Comment:
If we have multiple route groups, is the first one that fails going to throw an exception back to the client? How does this fail over to the "correct" route group?